### ISSUE TITLE
errorformat working with :make, but :MakeJob says "ended with 0 findings"

### ISSUE THREAD

**Issue by sampsyo** [djmoch/vim-makejob#7]
Along the lines of #5, I'm having trouble getting my `errorformat` settings to work with :MakeJob (and :LmakeJob) in the same way that they do with :make (and :lmake).

I'm running c5f304a of make-job via Plug. I've also defined a tiny little `.vim/compiler/flake8.vim` like this:

```vim
if exists('current_compiler')
    finish
endif
let current_compiler = 'flake8'

CompilerSet makeprg=flake8
CompilerSet errorformat=%f:%l:%c:\ %t%n\ %m
```

This works fine with `:make` and `:lmake`; errors are detected as you would expect. But running make-job just says "flake8 ended with 0 findings." Any ideas about where I should start debugging?

For completeness, here's the flake8 I'm using:

```
$ flake8 --version
3.2.1 (pyflakes: 1.3.0, mccabe: 0.5.2, pycodestyle: 2.2.0) CPython 2.7.12 on Darwin
```

**Comment by djmoch** [comment]
Hmm. I just used your compiler to run flake8 against a simple Python
file and everything worked as expected for both :make and :MakeJob.

Is there anything in the Quickfix window if you call :copen? MakeJob
only counts findings if errorformat is able to parse them. In other
words, if the errorformat doesn't parse the particular error your file
is generating, then the whole line will be read into Quickfix (without
any formatting), and MakeJob will NOT count it as a finding.

Another thing to consider is whether the preview window is catching any
output. You could put a breakpoint right at the top of s:JobHandler to
stop before the preview window closes and confirm.

Try out those suggestions and let me know what you find.

**Comment by sampsyo** [comment]
Thanks for the idea! The quickfix window revealed that the `flake8` invocation was crashing. I dug a little deeper and found that, while :make is invoking `flake8` with no arguments, :MakeJob is invoking the command with a single, empty-string argument. I'm not sure why this manifested on my machine but not yours, but I believe it is due to the space inserted on this line: https://github.com/djmoch/vim-makejob/blob/c5f304a3ff71451a569aa42deb3afad9859d7643/plugin/makejob.vim#L109

I thought I might work around this by adding a % to my `makeprg` line to include the current filename, which is probably what I want anyway. But it seems like the plugin doesn't currently interpolate % and # as the built-in make does. (Unless I'm missing something!)

**Comment by sampsyo** [comment]
(I can confirm that commenting out that `let make = make.' '.expand(a:1)` line does make the empty-string argument go away. At the expense of all the other arguments, of course.)

**Comment by djmoch** [comment]
Line 109 isn't the issue, because if you provide no arguments to
:MakeJob, then line 99 returns false and the whole block is skipped.
However, I was able to reproduce your error by adding a space after the
:MakeJob call with no other arguments (so ':MakeJob <CR>' instead of
':MakeJob<CR>'). Is it possible that's what's going on here? Either way,
I should probably check for arguments containing only whitespace.

The same basic issue held true with '%'. Line 109 tries to expand '%',
but I was able to break that functionality by adding a space after the
percent sign. So not only do I need to be checking for purely whitespace
arguments, but I need to be trimming supplied arguments as well.

I'll try to push an update sometime later today to address this. In the
meantime, let me know if I've located your issue. If I have, then the
simple workaround is to remove trailing spaces from your invocation of
:MakeJob.

**Comment by sampsyo** [comment]
Wow; weird! You're right that everything should work without trailing whitespace. But I never typed the space after :MakeJob, and trying again confirms that the plugin seems to "receive" a space in its arguments even when I don't type one. I can't immediately explain why that's happening, but I'll keep digging.

**Comment by djmoch** [comment]
See if 418b222e80369c3e6ac2bd593f4dfdbb49e6d999 fixes your issue.

**Comment by sampsyo** [comment]
Yes, it seems to work! At least half way—the plain `flake8` invocation now works as expected, but using `%` still doesn't get interpolated. I'll look into that more deeply also.

**Comment by djmoch** [comment]
Out of curiosity, if you break at the beginning of s:MakeJob and type
'echo a:1', what value do you see? For me, the '%' is being expanded
before ever entering the function. If that's not what you're seeing,
then there's probably a setting in Vim that controls that behavior that
I'm not aware of.

**Comment by sampsyo** [comment]
Ah—I meant that I used % in my `makeprg` setting, not as an argument to :MakeJob. You're absolutely right that typing `:MakePrg %` or somesuch does do the interpolation, as confirmed by echoing `a:1` at the top of the function.

**Comment by djmoch** [comment]
Oh, yeah including the percent sign in your `makeprg` is not going to work. That sort of expansion only happens for wildcards passed into the Ex command line directly.

What I do to avoid having to type the percent sign is include it in my mapping. So, something like: `nnoremap <F9> :MakeJob %<CR>`.

Is there anything else that still isn't working?

**Comment by sampsyo** [comment]
Got it; good call.

For what it's worth, the built-in :make does expand % and # in `makeprg` (c.f. [docs](http://vimdoc.sourceforge.net/htmldoc/options.html#'makeprg')). So, if it seems interesting, I might suggest is as yet one more way to make this plugin a "drop-in replacement" for stock :make.

Other than that: yes! Thank you for sorting this out (and for a great addition to vim). :sparkles:

**Comment by djmoch** [comment]
Look at that. I must have misunderstood that statement when I was reading the docs (and never thought to test it, since I obviously don't use it).

I'll go ahead and re-open this until I can implement.

**Comment by djmoch** [comment]
`expand(token)` can result in an empty string, which means bef91c6c585240031268e3de8c1b684fce8927df is broken for some values of `makeprg`. Reopening until I can fix this.

**Comment by djmoch** [comment]
After some fits and starts (especially in Windows), I'm satisfied this is working now. Fixed in 4b53b995085b12dc903fec82104b7701fec891e1.

**Comment by sampsyo** [comment]
Woohoo! This works exactly as I expect it to now. Thank you! :sparkles:
